### PR TITLE
Support 'include_type_name' in RestGetIndicesAction

### PR DESCRIPTION
**Pull request by cbuescher**
This is the backport of #37149, introducing the "include_type_name" parameter
to the indices.get API. The difference is that the parameter defaults to "true" on 6.x,
leading to the old known response format. A deprecation is issues when the parameter
isn't used or is set to "true" (the default) to make users aware of the changed default
behaviour starting in 7.0 and moving them towards the new typeless response format.

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml [hunk 1]
```diff
@@ -52,6 +52,35 @@ setup:
   - is_true: test_index.settings
   - is_true: test_index.mappings
 
+---
+"Test include_type_name":
+  - skip:
+      version: " - 6.6.99"
+      reason: the include_type_name parameter is not backported to pre 6.7 versions yet
```
Comment (jtibshirani) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: Small comment, maybe this should read 'the include_type_name parameter is not supported before 6.7', since we don't intend backport this further.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java [hunk 2]
```diff
@@ -41,6 +46,14 @@
  */
 public class RestGetIndicesAction extends BaseRestHandler {
 
+    private static final DeprecationLogger deprecationLogger = new DeprecationLogger(LogManager.getLogger(RestGetIndicesAction.class));
+    static final String TYPES_DEPRECATION_MESSAGE = "[types removal] The response format of get indices requests will change in "
+            + "the next major version. Please start using the `include_type_names` parameter set to `false` in the request to "
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java]: Small typo: `include_type_names` -> `include_type_name`.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java [hunk 3]
```diff
@@ -58,6 +71,9 @@ public String getName() {
     @Override
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
         String[] indices = Strings.splitStringByCommaToArray(request.param("index"));
+        if (request.hasParam(INCLUDE_TYPE_NAME_PARAMETER) == false || request.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER, true) == true) {
```
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java]: Could this just be `if (request.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER, true)) { ... }`?

File: server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+public class RestGetIndicesActionTests extends ESTestCase {
+
+    /**
+     * Test that setting no "include_type_name" or setting it to "true" raises a warning
+     */
+    public void testIncludeTypeNamesWarning() throws IOException {
+        Map<String, String> params = new HashMap<>();
+        if (randomBoolean()) {
+            params.put(INCLUDE_TYPE_NAME_PARAMETER, randomFrom("true"));
```
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesActionTests.java]: This can just be `params.put(INCLUDE_TYPE_NAME_PARAMETER, "true");`.

File: docs/reference/indices/get-index.asciidoc [hunk 5]
```diff
@@ -15,3 +15,64 @@ alias or wildcard expression is required.
 
 The get index API can also be applied to more than one index, or on
 all indices by using `_all` or `*` as index.
+
+[float]
+=== Skipping types
+
+Types are scheduled to be fully removed in Elasticsearch 8.0 and will not appear
```
Comment (jtibshirani) [inline docs/reference/indices/get-index.asciidoc]: I think it would be nice to mention that the type will not be returned in 7.0 by default. This is also what we do in other 6.x documentation: https://github.com/elastic/elasticsearch/commit/cd904337c6651578ab13ee4b1c74bdacb7b1a061#diff-0ee8c56315540fb266357d5163321ecf

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java [hunk 6]
```diff
@@ -117,11 +117,15 @@ public AutoFollowStats(StreamInput in) throws IOException {
         numberOfFailedRemoteClusterStateRequests = in.readVLong();
         numberOfSuccessfulFollowIndices = in.readVLong();
         if (in.getVersion().onOrAfter(Version.V_6_7_0)) {
-            recentAutoFollowErrors = new TreeMap<>(in.readMap(StreamInput::readString,
-                in1 -> new Tuple<>(in1.readZLong(), in1.readException())));
+            // need some explicit type casting here to infer arguments correctly in Eclipse IDE
+            Map<String, Tuple<Long, ElasticsearchException>> autoFollowErrors = in.readMap(StreamInput::readString,
```
Comment (jtibshirani) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java]: It seems nice to do this in a separate commit. Totally up to you though, I don't want to make a big hassle.

**Review by jtibshirani — COMMENTED** [review]
This is looking good to me, I just left some small comments.

**Comment by jtibshirani** [comment]
One other comment I had forgotten -- should we start explicitly setting `include_type_name = true` now in the Java HLRC, to make sure that the 6.7 client is compatible with 7.0, which has a different default for `include_type_name`?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml (hunk 1 above)
Comment (cbuescher) [reply rest-api-spec/src/main/resources/rest-api-spec/test/indices.get/10_basic.yml]: sure, easy to change

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java (hunk 2 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java]: Good one, I was thinking about using the constant itself at some point to avoid typos like this but then decided against it for better readability. Will correct this.

File: docs/reference/indices/get-index.asciidoc (hunk 5 above)
Comment (cbuescher) [reply docs/reference/indices/get-index.asciidoc]: I'll just take that text-blob then

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java (hunk 6 above)
Comment (cbuescher) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java]: Its already in another PR, this will go away when merging in master, I just needed it to work on this branch without compile errors from my IDE.

Comment (cbuescher) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ccr/AutoFollowStats.java]: Merging in 6.x I mean. Its 3863435a0da1824013214642971041fe8317883a

**Comment by cbuescher** [comment]
> Should we start explicitly setting `include_type_name = true` now in the Java HLRC, to make sure that the 6.7 client is compatible with 7.0

This should already happen by using the default value `true` if no parameter is set on the request. By testing this we should also catch accidental changes to the default, since the HLRC client will not be able to parse the response. I think that should be enough, no?

File: server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java [hunk 7]
```diff
@@ -249,15 +252,33 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                     }
                     builder.endObject();
 
-                    builder.startObject("mappings");
                     ImmutableOpenMap<String, MappingMetaData> indexMappings = mappings.get(index);
-                    if (indexMappings != null) {
+                    // the default on 6.x should be true to include types in the response
+                    boolean includeTypeName = params.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER, true);
```
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: I now have this default setting define in the constant BaseRestHandler.DEFAULT_INCLUDE_TYPE_NAME_POLICY

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/admin/indices/get/GetIndexResponse.java]: Thanks

File: rest-api-spec/src/main/resources/rest-api-spec/api/indices.get.json [hunk 8]
```diff
@@ -13,6 +13,10 @@
         }
       },
       "params":{
+        "include_type_name": {
+          "type" : "boolean",
+          "description" : "Whether to add the type name to the response (default: false)"
```
Comment (markharwood) [inline rest-api-spec/src/main/resources/rest-api-spec/api/indices.get.json]: default should be true?

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java [hunk 9]
```diff
@@ -58,6 +71,9 @@ public String getName() {
     @Override
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
         String[] indices = Strings.splitStringByCommaToArray(request.param("index"));
+        if (request.paramAsBoolean(INCLUDE_TYPE_NAME_PARAMETER, true)) {
```
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetIndicesAction.java]: Another candidate for referring to BaseRestHandler.DEFAULT_INCLUDE_TYPE_NAME_POLICY but no big deal if you miss this

File: server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java [hunk 10]
```diff
@@ -83,22 +85,27 @@ protected GetMappingsResponse mutateInstance(GetMappingsResponse instance) throw
     public static ImmutableOpenMap<String, MappingMetaData> createMappingsForIndex() {
         // rarely have no types
         int typeCount = rarely() ? 0 : scaledRandomIntBetween(1, 3);
+        return createMappingsForIndex(typeCount);
```
Comment (markharwood) [inline server/src/test/java/org/elasticsearch/action/admin/indices/mapping/get/GetMappingsResponseTests.java]: I'm not sure I understand when/why this would work? I can understand zero or one mappings for an index but shouldn't >1 be impossible/an error?

**Comment by jtibshirani** [comment]
> This should already happen by using the default value true if no parameter is set on the request.

We would like the 6.7 HLRC to be able to understand responses from 7.0 nodes. Since `include_type_name` will default to `false` in 7.0, we therefore need to explicitly set `include_type_name=true` when creating the HLRC request.

File: client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesRequestConvertersTests.java [hunk 11]
```diff
@@ -364,6 +365,8 @@ public void testGetIndex() throws IOException {
         RequestConvertersTests.setRandomIndicesOptions(getIndexRequest::indicesOptions, getIndexRequest::indicesOptions, expectedParams);
         RequestConvertersTests.setRandomLocal(getIndexRequest, expectedParams);
         RequestConvertersTests.setRandomHumanReadable(getIndexRequest, expectedParams);
+        // Force "include_type_name" parameter since responses need to be compatible when coming from 7.0 nodes
```
Comment (jtibshirani) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/IndicesRequestConvertersTests.java]: Maybe this comment should go on `IndicesRequestConverters`?

**Review by jtibshirani — APPROVED** [review]
Looks good to me, pending @markharwood's last small comments.

**Comment by markharwood** [comment]
LGTM

**Comment by cbuescher** [comment]
@jtibshirani @markharwood thanks for the reviews. I'll merge and follow up with a small PR against master enabling mixed-cluster rest tests between 7.0 and 6.x now that the parameter is supported in both versions.
